packages/pulseaudio/module-sles-source.c

```diff
@@ -21,338 +21,377 @@
+            sl->formatType     = SL_ANDROID_DATAFORMAT_PCM_EX;
+            sl->endianness     = SL_BYTEORDER_LITTLEENDIAN;
+            sl->representation = SL_ANDROID_PCM_REPRESENTATION_FLOAT;
+            break;
+        case PA_SAMPLE_FLOAT32BE:
+            sl->formatType     = SL_ANDROID_DATAFORMAT_PCM_EX;
+            sl->endianness     = SL_BYTEORDER_BIGENDIAN;
+            sl->representation = SL_ANDROID_PCM_REPRESENTATION_FLOAT;
+            break;
+        default:
+            pa_log_error("Unsupported sample format: only s16/24/32 and f32 LE/BE are supported.");
+            return 1;
+    }
+    sl->bitsPerSample = sl->containerSize = pa_sample_size(ss) * 8;
+
+    return 0;
+}
 
 static int pa_init_sles_record(struct userdata *u, pa_sample_spec *ss) {
-    pa_assert(u); 
-    SLboolean	 required[2];
-    SLInterfaceID  iidArray[2];
-   
-    required[0] = SL_BOOLEAN_FALSE;
-    iidArray[0] = SL_IID_ANDROIDSIMPLEBUFFERQUEUE;
-    required[1] = SL_BOOLEAN_FALSE;
-    iidArray[1] = SL_IID_ANDROIDCONFIGURATION;
-
-    SLAndroidConfigurationItf configItf;    
-    SLDataSource           recSource;
-    SLDataLocator_IODevice ioDevice = {SL_DATALOCATOR_IODEVICE,
-    SL_IODEVICE_AUDIOINPUT,
-    SL_DEFAULTDEVICEID_AUDIOINPUT, NULL};
-    SLDataSink                recDest;
-    SLDataLocator_AndroidSimpleBufferQueue recBuffQueue;
-    
-    ioDevice.locatorType = SL_DATALOCATOR_IODEVICE; 
-    ioDevice.deviceType = SL_IODEVICE_AUDIOINPUT; 
-    ioDevice.deviceID = SL_DEFAULTDEVICEID_AUDIOINPUT;
-    ioDevice.device = NULL;
-    recSource.pLocator = (void *) &ioDevice; 
-    recBuffQueue.locatorType = SL_DATALOCATOR_ANDROIDSIMPLEBUFFERQUEUE;
-    recBuffQueue.numBuffers = 8; 
+    pa_assert(u);
+
+    // check and convert the sample spec
     SLAndroidDataFormat_PCM_EX pcm;
-    
-    if (ss->format == PA_SAMPLE_FLOAT32LE) {
-        pcm.formatType = SL_ANDROID_DATAFORMAT_PCM_EX;
-        pcm.representation = SL_ANDROID_PCM_REPRESENTATION_FLOAT;
-    } else {
-        pcm.formatType = SL_DATAFORMAT_PCM;
+    if (pa_sample_spec_to_sl_format(ss, &pcm))
+        return 1;
+
+    // common sles error handling
+    #define CHK(stmt) {                                                        \
+        SLresult res = (stmt);                                                 \
+        if (res != SL_RESULT_SUCCESS) {                                        \
+            pa_log_error("Failed to initialize OpenSL ES: error %d at %s:%d",  \
+                res, __FILE__, __LINE__);                                      \
+            return 1;                                                          \
+        }                                                                      \
     }
-    
-    pcm.numChannels = ss->channels;
-    pcm.sampleRate = ss->rate * 1000;
-    pcm.bitsPerSample = pcm.containerSize = pa_sample_size(ss) * 8;
-    pcm.channelMask = SL_SPEAKER_FRONT_LEFT;
-    pcm.endianness = SL_BYTEORDER_LITTLEENDIAN; 
```

Comment:
For this you can use:
https://github.com/pulseaudio/pulseaudio/blob/v14.2/src/pulse/sample.c#L255

Reply:
Fixed.